### PR TITLE
Fix issues from #1047

### PR DESCRIPTION
**Pull request by kevinbarabash**
## Summary:
I didn't put the 'waiting' logic for the server-side nav case in the right location.  This PR fixes that.

One of the things from the ADR was wanted to make sure people are using 'beforeNav' and 'safeWithNav' only when 'href' is set.  This PR also fixes that by changing the Button props to be a disjoint union.

Issue: WEB-2990

## Test plan:
- yarn jest

### CODE REVIEW

**Review by jeresig — APPROVED** [review]

File: packages/wonder-blocks-button/components/__tests__/button.flowtest.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import * as React from "react";
+
+import Button from "../button.js";
+
+// $ExpectError: href must be used with beforeNav
+<Button beforeNav={() => Promise.resolve()}>Hello, world!</Button>;
+
+// $ExpectError: href must be used with safeWithNav
+<Button safeWithNav={() => Promise.resolve()}>Hello, world!</Button>;
+
```
Comment (aag) [inline packages/wonder-blocks-button/components/__tests__/button.flowtest.js:12]: Maybe add one more test to verify that `href` can be used by itself?

```
// It's okay to use href by itself
<Button href="/foo">
    Hello, world!
</Button>;
```

**Review by aag — APPROVED** [review]
Looks good to me. I suggested 1 additional test, but I'll leave it up to you to add it or not.

File: packages/wonder-blocks-button/components/button.js [hunk 2]
```diff
@@ -134,25 +134,35 @@ export type SharedProps = {|
...
+
+          /**
+           * Run async code in the background while client-side navigating. If the
+           * navigation is server-side, the callback must be settled before the
+           * navigation will occur. Errors are ignored so that navigation is
+           * guaranteed to succeed.
+           */
+          safeWithNav?: () => Promise<mixed>,
+      |};
+
```
Comment (somewhatabstract) [inline packages/wonder-blocks-button/components/button.js:165]: Thanks for doing this. Be prepared for webapp issues though if folks spread these props for some reason. Our version of flow in webapp is not good to spread unions.

File: packages/wonder-blocks-core/components/clickable-behavior.js [hunk 3]
```diff
@@ -329,15 +329,16 @@ export default class ClickableBehavior extends React.Component<
...
-                        // indicating that we're waiting for navigation to occur.
-                        this.setState({waiting: true});
-                    }
+                    // client-side nav
                     safeWithNav();
                 } else {
                     try {
+                        if (!this.state.waiting) {
+                            // We only show the spinner for safeWithNav when doing
+                            // a server-side navigation since since the spinner is
```
Comment (somewhatabstract) [inline packages/wonder-blocks-core/components/clickable-behavior.js:338]: This terminology is a bit confusing (at least for me) as I immediately think of "server-side rendering". Perhaps consider "full page load navigation"?

**Review by somewhatabstract — APPROVED** [review]
LGTM

Some inline thoughts. Nothing serious.

File: packages/wonder-blocks-button/components/button.js (hunk 2 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-button/components/button.js:165]: Thanks for the heads up.

File: packages/wonder-blocks-core/components/clickable-behavior.js (hunk 3 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-core/components/clickable-behavior.js:338]: "full page load navigation" is better.  I'll update this before landing.

**Pull request merged**